src/emlid_reach.cpp

```diff
@@ -0,0 +1,434 @@
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file emlid_reach.cpp
+ *
+ * @author Bastien Auneau <bastien.auneau@while-true.fr>
+ */
+
+
+#include "emlid_reach.h"
+
+#include <stdio.h>
+#include <math.h>
+#include <string.h>
+#include <ctime>
+#include <cmath>
+#include <cstring>
+#include <errno.h>
+
+#include <stdlib.h>
+
+//// ERB
+// 'E' Ox45 | 'R' Ox52 | ID | LENGTH (2 Bytes little endian) | payload | CHECKSUM (2B)
+#define ERB_SYNC_1 0x45 // E
+#define ERB_SYNC_2 0x52 // R
+
+#define ERB_ID_VERSION         0x01
+#define ERB_ID_GEODIC_POSITION 0x02
+#define ERB_ID_NAV_STATUS      0x03
+#define ERB_ID_DOPS            0x04
+#define ERB_ID_VELOCITY_NED    0x05
+#define ERB_ID_SPACE_INFO      0x06 // not used, reduces stack usage
+#define ERB_ID_RTK             0x07 // RTK, TODO
+
+
+#define EMLID_UNUSED(x) (void)x;
+
+#define GPS_PI 3.141592653589793238462643383280f
+
+#define AUTO_DETECT_MAX_TIMEOUT       10 // if more detect failed AND
+#define AUTO_DETECT_MAX_PARSE_ERR     2  // if more detect failed AND
+#define AUTO_DETECT_MAX_READ_SENTENCE 10 // if more detect succeed
+
+
+GPSDriverEmlidReach::GPSDriverEmlidReach(GPSCallbackPtr callback, void *callback_user,
+		struct vehicle_gps_position_s *gps_position, struct satellite_info_s *satellite_info) :
+	GPSHelper(callback, callback_user),
+	_gps_position(gps_position), _satellite_info(satellite_info)
+{}
+
+
+int
+GPSDriverEmlidReach::configure(unsigned &baudrate, OutputMode output_mode)
+{
+	// TODO RTK
+	if (output_mode != OutputMode::GPS) {
+		GPS_WARN("EMLIDREACH: Unsupported Output Mode %i", (int)output_mode);
+		return -1;
+	}
+
+	unsigned baud_allowed[] {57600, 115200, 230400};
+
+	for (unsigned i = 0; i < sizeof(baud_allowed) / sizeof(baud_allowed[0]); i++) {
+		if (baudrate > 0 && baudrate != baud_allowed[i]) {
+			continue;
+		}
+
+		_sentence_cnt = 0;
+		_parse_err_cnt = 0;
+		_erb_decode_state  = ERB_State::init;
+		_read_buff_len = 0;
+		_checksum_buff_len = 0;
+		_read_buff_ptr = _read_buff + GPS_READ_BUFFER_SIZE;
+		_erb_payload_len = 0;
+
+		if (GPSHelper::setBaudrate(baud_allowed[i]) != 0) {
+			continue;
+		}
+
+		if (! testConnection()) {
+			continue;
+		}
+
+		GPS_INFO("EmlidReach: config OK with ERB baudrate: %d", baud_allowed[i]);
+		baudrate = baud_allowed[i];
+		return 0;
+	}
+
+	return -1;
+}
+
+
+bool
+GPSDriverEmlidReach::testConnection()
+{
+	_testing_connection = true;
+
+	unsigned timeout_cnt = 0;
+
+	while (timeout_cnt < AUTO_DETECT_MAX_TIMEOUT
+	       && _parse_err_cnt < AUTO_DETECT_MAX_PARSE_ERR
+	       && _sentence_cnt < AUTO_DETECT_MAX_READ_SENTENCE) {
+		if (receive(50000) == 0b100) { // timeout larger than what defined in read() are truncated
```

Comment:
Please don't rely on a truncated timeout and use something like 50 or 100 ms here.
I'd also rather check if there is one (or more) successfully parsed message, instead of checking for parser errors.

Reply:
timeout set to 100msec
testConnection() returns true on number of successfully parsed message